logging/net_appender.go

```diff
@@ -223,19 +234,27 @@ func (nl *NetAppender) syncOnce() (bool, error) {
 		batchSize = len(nl.toLog)
 	}
 
-	// Pull a batch from the queue, unlock mutex, and return an error if write
-	// fails. We are ok losing logs (pulling a batch out of toLog even in the
-	// event of a write failure) in order to not block adding to the queue while
-	// writing.
+	// Read a batch from the queue, unlock mutex, and return an error if write
+	// fails. Lock mutex again to remove batch from queue only if write succeeded
+	// and front of queue was not mutated by addToQueue/addBatchToQueue throwing
+	// away the oldest logs due to overflows beyond maxQueueSize.
 	batch := nl.toLog[:batchSize]
-	nl.toLog = nl.toLog[batchSize:]
-	moreWorkToDo := len(nl.toLog) > 0
 	nl.toLogMutex.Unlock()
 
 	if err := nl.remoteWriter.write(batch); err != nil {
 		return false, err
 	}
-	return moreWorkToDo, nil
+
+	nl.toLogMutex.Lock()
+	defer nl.toLogMutex.Unlock()
+
+	// If we've overflowed more times than the size of the batch we wrote, do not
+	// mutate toLog at all.
```

Comment:
```suggestion
	// Remove successfully synced logs from the queue. If we've overflowed more times than the size of the batch
	//  we wrote, do not mutate toLog at all.
```
Added another sentence before to better describe the if